### PR TITLE
Add Sidecar front IO HSC control, work around front IO PHY oscillator errata

### PR DESCRIPTION
**Pull request by arjenroodselaar**
The oscillator used to drive the VSC8562 PHY on the Sidecar front IO board occasionally fails to start properly, causing the QSGMII link to the management network switch to fail and the technician ports to be non-functional (#1410). This diff does the following:

- Implement sequencer control over the front IO hot swap controller
- Make monorail-server detect the oscillator failure, request a power cycle of the front IO board and retry PHY initialization until the QSGMII link is operational
- Refactor power rail status primitives into a new `fpga-app-api` crate

### CODE REVIEW

File: drv/fpga-app-api/Cargo.toml
Comment (mkeeter) [inline drv/fpga-app-api/Cargo.toml:6]: Nit: you can delete this and the empty `[build-dependencies]` below

File: drv/sidecar-front-io/src/controller.rs
Comment (mkeeter) [inline drv/sidecar-front-io/src/controller.rs:32]: Nit:
```suggestion
        self.fpga.state()? == DeviceState::RunningUserDesign
```

File: drv/sidecar-seq-server/src/front_io.rs
Comment (mkeeter) [inline drv/sidecar-seq-server/src/front_io.rs:11]: Why remove `fruid` and `i2c_task` here?  (they're here as a small caching optimization, so we don't need to look them up repeatedly)`

File: drv/sidecar-seq-server/src/front_io.rs
Comment (mkeeter) [inline drv/sidecar-seq-server/src/front_io.rs:45]:
```suggestion
        Ok(self.controllers.iter().all(|c| c.ready().unwrap_or(false)))
```
(also, is it okay to discard errors from `ready()` here?)

File: drv/sidecar-front-io/src/phy_smi.rs
Comment (mkeeter) [inline drv/sidecar-front-io/src/phy_smi.rs:96]:
```suggestion
        let good = phy_osc & Reg::VSC8562::PHY_OSC::GOOD != 0;
        let valid = phy_osc & Reg::VSC8562::PHY_OSC::VALID != 0;
        Ok(Some(good).filter(|_| valid))
```

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (mkeeter) [inline task/monorail-server/src/bsp/sidecar_bc.rs:155]: Since `front_io_board_present` is infallible on the `sidecar-seq` side, we could change the reply type in Idol to `Simple(bool)`.

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (mkeeter) [inline task/monorail-server/src/bsp/sidecar_bc.rs:396]:
```suggestion
            // Request a reset of the PHY.  If we had previously marked the PHY
            // oscillator as bad, then this power-cycles the entire front IO
            // board; otherwise, it only power-cycles the PHY.
```

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (mkeeter) [inline task/monorail-server/src/bsp/sidecar_bc.rs:398]: It seems a little silly that we smuggle the `OSC_GOOD/VALID` bits _through_ the FPGA into the `sidecar-seq` task.

We could instead track oscillator good/valid in this task and add a parameter to the RPC function, e.g. `fn reset_front_io_phy(full_power_cycle: bool)`.

On the other hand, we need to put those bits in the FPGA for other reasons, so maybe it's fine 🤷🏻‍♂️

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (mkeeter) [inline task/monorail-server/src/bsp/sidecar_bc.rs:235]: I'd be tempted to move this into a `fn is_front_io_link_good(...)`, just for clarity.

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (mkeeter) [inline task/monorail-server/src/bsp/sidecar_bc.rs:230]: This loop should only run if `self.vsc8562.is_some()`; otherwise, we'll get stuck here forever if the front IO board is missing.

File: drv/fpga-app-api/src/power_rail.rs
Comment (mkeeter) [inline drv/fpga-app-api/src/power_rail.rs:26]: Since this doesn't actually validate that the `u8` is valid, I'd suggest calling it `RawPowerRailState` instead.

File: drv/sidecar-front-io/sidecar_qsfp_x32_controller_rev_b_c.bit
Comment (mkeeter) [inline drv/sidecar-front-io/sidecar_qsfp_x32_controller_rev_b_c.bit:1]: Before merging, we should update the README, etc to point at the new buildomat job.

File: drv/sidecar-mainboard-controller/src/front_io.rs
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/front_io.rs:41]: There's an `impl From<bool> for WriteOp`, so this could be
```suggestion
        self.fpga.write(
            enable.into(),
            Addr::FRONT_IO_STATE,
            Reg::FRONT_IO_STATE::ENABLE,
        )
```

File: drv/sidecar-mainboard-controller/src/tofino2.rs
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/tofino2.rs:159]:
```suggestion
            id: TofinoPowerRailId::from_usize(data.0)
```
seems to work

File: drv/sidecar-mainboard-controller/src/tofino2.rs
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/tofino2.rs:157]: I'm not convinced that `impl TryFrom` is better than
```rust
impl TofinoPowerRail {
    fn new(id: usize, state: PowerRailState) -> Result<Self, FpgaError> {
        // ...
    }
}
```
since constructing / indexing the tuple is awkward.

File: drv/fpga-app-api/src/lib.rs
Comment (mkeeter) [inline drv/fpga-app-api/src/lib.rs:9]: Why put this in a separate module, versus the existing `fpga-api`?  Are there more pieces of `fpga-api` that should move here eventually?

File: drv/fpga-app-api/src/power_rail.rs
Comment (mkeeter) [inline drv/fpga-app-api/src/power_rail.rs:89]: I have a weak preference for explicit shifts, e.g.
```suggestion
            enable: state.0 & (1 << 0) != 0,
            good: state.0 & (1 << 1) != 0,
            fault: state.0 & (1 << 2) != 0,
            vrhot: state.0 & (1 << 3) != 0,
```
(too bad the columns don't line up)

File: drv/sidecar-front-io/sidecar_qsfp_x32_controller_regs.json
Comment (mkeeter) [reply drv/sidecar-front-io/sidecar_qsfp_x32_controller_regs.json:428]: Okay, we're actually using this field to (persistently) pass data from one task to the other, which is a little weird but probably okay.

File: drv/sidecar-front-io/src/phy_smi.rs
Comment (mkeeter) [inline drv/sidecar-front-io/src/phy_smi.rs:48]: Why rename these?  I'd be worried about potential confusion with front IO HSC control; the old name is nicely explicit that this is PHY power control.

File: drv/fpga-app-api/src/power_rail.rs
Comment (Aaron-Hartwig) [inline drv/fpga-app-api/src/power_rail.rs:12]: We have a pattern to tightly integrate types defined via RDL into our BSV/Rust code. At first blush, leaning on that could easily eliminate the duplicate hand coding of the `PowerRailStatus` enum. I don't recall if we support generating structs off hand, so `PowerFailPinState` may not be covered but that seems like a natural place to continue to expand our integration here lest we need to change RDL, BSV, and Rust every time a register changes.

File: drv/sidecar-front-io/src/phy_smi.rs
Comment (Aaron-Hartwig) [reply drv/sidecar-front-io/src/phy_smi.rs:48]: I'm inclined to agree with Matt here, but presumably this was done since the functions are within the `phy_smi` module so adding `phy` into all the names is a bit redundant. I tend to have more verbose tendencies with naming though.

File: drv/sidecar-front-io/sidecar_qsfp_x32_controller_regs.json
Comment (arjenroodselaar) [reply drv/sidecar-front-io/sidecar_qsfp_x32_controller_regs.json:428]: Yes, let me elaborate on this a bit. One of my design goals for the sequencer task is that it (or the SP as a whole for that matter) can safely be restarted without interrupting Tofino once the main- and front IO board controllers have been initialized and configured. Hence the external thermal sensor, sequencer logic in the controller etc. There are caveats here that the bitstream versions of those controllers don't change, the clock generator configuration does not change etc., but many times it should be possible to leave Tofino running.

In order to stay true to this goal I needed a more persistent place than sequencer RAM to store the result of the question whether or not the PHY oscillator came out of reset correctly on the last board power up. Otherwise it would either have to take a guess or proactively power cycle the front IO board, interrupting the transceivers. Storing this bit of information with the front IO controller which manages the PHY felt like a reasonable thing to do since it'll automatically get reset if power to the board is cycled, the bitstream of the controller is reloaded or the design is reset (in which case I think the PHY gets reset too).

We're not actually using it to pass data between tasks though. `monorail-server` tells the sequencer that it found the PHY functional, at which point we assume the oscillator is good, and the sequencer then stores it in the front IO board controller (using an API from the `sidecar-front-io-board` crate).

File: drv/sidecar-mainboard-controller/src/tofino2.rs
Comment (arjenroodselaar) [reply drv/sidecar-mainboard-controller/src/tofino2.rs:295]: Ha, nice! Thanks.

File: drv/sidecar-seq-server/src/front_io.rs
Comment (arjenroodselaar) [reply drv/sidecar-seq-server/src/front_io.rs:11]: Because they're only used in the `FrontIOBoard::present(..)` function which, unless a front IO board power cycle is requested, is only called once on task init (through `ServerImpl::front_io_board_preinit`). I figured I'd let the lookup happen then rather than on struct creation. And when the `fruid` handle is used we're interacting with the EEPROM over I2C, which I figured dwarfs any speedups we'd get from keeping this handle around.

But I did realize that during sequencer `main` we do get the task handle repeatedly, so there it makes sense to do it once and use it several times.

File: drv/sidecar-seq-server/src/front_io.rs
Comment (arjenroodselaar) [reply drv/sidecar-seq-server/src/front_io.rs:45]: Ah nice, thank you.

The function is used in two places; 1) determine of the bitstream may need to be loaded in the controllers, which performs a more detailed check, and 2) to gate the `transceiver` task and the sequencing decision to start Tofino until the sequencer is sure a future power cycle of the front IO board is not needed. In both cases we're really only interested in the happy answer and take the conservative path (default to false) otherwise.

I realized the function signature does not reflect this so it now returns a `bool` instead of a `Result`.

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (arjenroodselaar) [reply task/monorail-server/src/bsp/sidecar_bc.rs:155]: TIL! That looks much better indeed.

File: drv/fpga-app-api/src/lib.rs
Comment (arjenroodselaar) [reply drv/fpga-app-api/src/lib.rs:9]: The intend is that `fpga-api` deals with FPGA devices (reset, loading bitstreams etc.) and `fpga-app-api` would be APIs for logic contained within bitstreams and running on those FPGAs. An alternative name for this crate could be `fpga-design-api`. The point is that there is RTL we're starting to share between multiple tasks and I figured I'd start pulling some of that into it's own crate. Happy to hear suggestions.

File: drv/sidecar-front-io/sidecar_qsfp_x32_controller_rev_b_c.bit
Comment (arjenroodselaar) [reply drv/sidecar-front-io/sidecar_qsfp_x32_controller_rev_b_c.bit:1]: Definitely. The logic pieces still need to land in quartz/main after which I can pull the final artifacts from the resulting job.

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (arjenroodselaar) [reply task/monorail-server/src/bsp/sidecar_bc.rs:398]: I should have done a better job documenting this :) Please see https://github.com/oxidecomputer/hubris/pull/1449#discussion_r1253963119. Yes, `monorail-server` does technically have access to this register (because it needs access to the adjacent SMI registers), the expectation is that the sequencer maintains this state (and updates that register).

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (arjenroodselaar) [reply task/monorail-server/src/bsp/sidecar_bc.rs:230]: Ah yes, thank you!

File: drv/sidecar-mainboard-controller/src/front_io.rs
Comment (arjenroodselaar) [reply drv/sidecar-mainboard-controller/src/front_io.rs:41]: Heh yes, I am already forgetting what I wrote less than 6 months ago :/

File: drv/fpga-app-api/src/power_rail.rs
Comment (arjenroodselaar) [reply drv/fpga-app-api/src/power_rail.rs:89]: Fair, this does aid readability because you can read off the bit index.

File: drv/sidecar-front-io/src/phy_smi.rs
Comment (arjenroodselaar) [reply drv/sidecar-front-io/src/phy_smi.rs:48]: Yeah I am all about localized naming. But if both of you like having the phy portion in there I'll put it back.

File: drv/fpga-app-api/src/power_rail.rs
Comment (arjenroodselaar) [reply drv/fpga-app-api/src/power_rail.rs:12]: Yes, the reason this looks the way it does is because while RDL can define types, we currently can't export those types through the JSON representation into Hubris. We can only generate specific instantiations of that type within a register map. We should figure out how to extend our capabilities here, but I don't want to hold up this diff on something open-ended.

File: drv/sidecar-mainboard-controller/src/tofino2.rs
Comment (arjenroodselaar) [reply drv/sidecar-mainboard-controller/src/tofino2.rs:157]: I am hesitant here. I understand what you're saying, but conceptually I believe that a `new(..)` function should not error. This may be a left over from my C++ days where throwing exceptions from a constructor was problematic, but I still like to separate memory allocation/initialization from functions which may cause errors. As such I feel like the `try_from(..)`, despite its awkward definition better conveys that we're doing something which has a change of failing. Is there a potential way/trait/pattern that may satisfy both objectives?

File: drv/fpga-app-api/src/lib.rs
Comment (mkeeter) [reply drv/fpga-app-api/src/lib.rs:9]: Maybe `fpga-user-api`, since we use "user design" to refer to the running code?

File: drv/sidecar-seq-server/src/front_io.rs
Comment (mkeeter) [reply drv/sidecar-seq-server/src/front_io.rs:11]: Gotcha, I don't feel too strongly about this

File: task/monorail-server/src/bsp/sidecar_bc.rs
Comment (mkeeter) [reply task/monorail-server/src/bsp/sidecar_bc.rs:391]: 👍🏻  seems fine to me

File: drv/sidecar-mainboard-controller/src/tofino2.rs
Comment (mkeeter) [reply drv/sidecar-mainboard-controller/src/tofino2.rs]: Having `new(...)` be fallible isn't _totally_ unheard of, e.g. [`NonZeroU8::new`](https://doc.rust-lang.org/std/num/struct.NonZeroU8.html#method.new) returns an `Option<Self>`.

It does appear to be uncommon, though.  What about making an infallible `new`
```rust
impl TofinoPowerRail {
    fn new(id: TofinoPowerRailId, status: PowerRailStatus, pins: PowerRailPinState) -> Self {
        // ...
    }
}
```
then doing the type conversions in the loop at [`tofino2.rs:292`](https://github.com/oxidecomputer/hubris/pull/1449/files#diff-580ceb53b19ee484f37faf79894c68765ef0a5ed11d39fa7b9d50641f2c3ce41R293)?  That appears to be the only place `try_from` is used.

Comment (arjenroodselaar) [reply drv/sidecar-mainboard-controller/src/tofino2.rs]: Alright, fair point. This only used in one place so we may as well pull the loop into that function. There's no need really to implement this in a separate `TryFrom`.

**Review by mkeeter — APPROVED** [review]

**Pull request merged**
